master/internal/stream/experiments.go

```diff
@@ -0,0 +1,277 @@
+package stream
+
+import (
+	"context"
+	"database/sql"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/uptrace/bun"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/stream"
+)
+
+const (
+	// ExperimentsDeleteKey specifies the key for delete experiments.
+	ExperimentsDeleteKey = "experiments_deleted"
+	// ExperimentsUpsertKey specifies the key for upsert experiments.
+	ExperimentsUpsertKey = "experiment"
+)
+
+// ExperimentMsg is a stream.Msg.
+// determined:streamable
+type ExperimentMsg struct {
+	bun.BaseModel `bun:"table:experiments"`
+	// immutable attributes
+	ID        int           `db:"id" json:"id"`
+	JobID     model.JobID   `db:"job_id" json:"job_id"`
+	OwnerID   *model.UserID `db:"owner_id" json:"owner_id"`
+	ProjectID int           `db:"project_id" json:"project_id"`
+	Unmanaged bool          `db:"unmanaged" json:"unmanaged"`
+
+	// mutable attributes
+	State     model.State `db:"state" json:"state"`
+	Archived  bool        `db:"archived" json:"archived"`
+	Progress  *float64    `db:"progress" json:"progress"`
+	StartTime time.Time   `db:"start_time" json:"start_time"`
+	EndTime   *time.Time  `db:"end_time" json:"end_time"`
+	Notes     string      `db:"notes" json:"notes"`
+
+	// metadata
+	Seq int64 `bun:"seq" json:"seq"`
+
+	// permission scope
+	WorkspaceID int `json:"-"`
+}
+
+// SeqNum returns the sequence number of an experiment message.
+func (e *ExperimentMsg) SeqNum() int64 {
+	return e.Seq
+}
+
+// UpsertMsg creates an Experiment upserted prepared message.
+func (e *ExperimentMsg) UpsertMsg() stream.UpsertMsg {
+	return stream.UpsertMsg{
+		JSONKey: ExperimentsUpsertKey,
+		Msg:     e,
+	}
+}
+
+// DeleteMsg creates an Experiment deleted prepared message.
+func (e *ExperimentMsg) DeleteMsg() stream.DeleteMsg {
+	deleted := strconv.FormatInt(int64(e.ID), 10)
+	return stream.DeleteMsg{
+		Key:     ExperimentsDeleteKey,
+		Deleted: deleted,
+	}
+}
+
+// ExperimentSubscriptionSpec is what a user submits to define an experiment subscription.
+// determined:streamable
+type ExperimentSubscriptionSpec struct {
+	ExperimentIds []int `json:"experiment_ids"`
+	Since         int64 `json:"since"`
+}
+
+func getExperimentMsgsWithWorkspaceID(expMsgs []*ExperimentMsg) *bun.SelectQuery {
+	q := db.Bun().NewSelect().Model(&expMsgs).
```

Comment:
question (non-blocking & out of scope): I'm getting concerned about the maintainability of this pattern. It'll likely result in funky bugs if new fields get added the <X>Msg struct, but not to this query. And requiring both to be updated seems redundant.

This is out of the scope for this PR (and this project), but I'll make a backlog to improve this function so that we have something that will easier to maintain.

Do you have any suggestions on if/how this could be improved? Maybe two separate queries that get joined and then executed?

Reply:
yeah I like the two separate queries idea - I feel like the pattern should be we use the `Msg` as the model and set it up to get all the fields defined in the struct. Then we can tack on a join as part of a separate callable function.

Reply:
The only semi-question I have is how bun would handle the fields not labeled with `bun:` - I would think (and hope) that it just ignores those fields